### PR TITLE
Support Google fonts other than Poppins

### PR DESCRIPTION
**Pull request by aashishgurung**
#### 1. Objective

Support Google fonts other than Poppins

#### 2. Description of change

Add `others` as an option to font name. This will activate a field where you can specify the Google font that you want to use

**🔧 Environments:**

- **WooCommerce**: v8.2.2
- **WordPress**: v6.4.1-alpha-57080
- **PHP version**: 8.1

### CODE REVIEW

**Review by muthuswamyopn — DISMISSED** [review]

File: includes/admin/views/omise-page-card-form-customization.php [hunk 1]
```diff
@@ -24,14 +24,23 @@
       <tr>
         <td class="text-bold" style="width: 250px;">Font Name</td>
         <td>
-          <select class="select-input" name="font[name]">
+          <select id="custom_sf_font_name" class="select-input" name="font[name]">
             <option value="Poppins">Poppins</option>
             <option value="Circular" selected>Circular</option>
+            <option value="Others" selected>Others</option>
```
Comment (danfowler) [inline includes/admin/views/omise-page-card-form-customization.php:30]: I think "Other" is preferable. It's shorter and means the same thing.

File: assets/javascripts/card-form-customization.js [hunk 2]
```diff
@@ -58,6 +58,24 @@ function getDesignFormValues() {
   return formValues;
 }
 
+function handleFontChange() {
+  const fontName = document.getElementById('custom_sf_font_name');
+
+  if (fontName.value == 'Others') {
```
Comment (danfowler) [inline assets/javascripts/card-form-customization.js:64]: 1. Can we define the value representing other font names as a constant?
2. Shall we always use strict equality?

File: assets/javascripts/card-form-customization.js [hunk 3]
```diff
@@ -58,6 +58,24 @@ function getDesignFormValues() {
...
+    document.getElementById('custom_sf_custom_font_name').style.display = null
+  }
+
+  fontName.addEventListener('change', (event) => {
+    if (event.target.value == 'Others') {
+      document.getElementById('custom_sf_custom_font_name').style.display = null
+    } else {
+      const customFontName = document.getElementById('custom_sf_custom_font_name');
+      customFontName.value = "";
+      customFontName.style.display = "none";
```
Comment (danfowler) [inline assets/javascripts/card-form-customization.js:74]: In this function, we're using both single and double quotes.

File: assets/javascripts/omise-embedded-card.js [hunk 4]
```diff
@@ -25,6 +25,12 @@ function showOmiseEmbeddedCardForm({
   }
   element.style.height = iframeElementHeight + 'px'
 
+  let fontName = font.name
+
+  if (font.name.trim() === "Others" && font.custom_name.trim() != "") {
```
Comment (danfowler) [inline assets/javascripts/omise-embedded-card.js:30]: There is also strict inequality `!==`

File: assets/javascripts/card-form-customization.js [hunk 5]
```diff
@@ -91,3 +109,4 @@ document.getElementById('omise-modal').addEventListener('click', (event) => {
 
 setDesignFormValues();
 handleColorInputChanges();
+handleFontChange();
```
Comment (danfowler) [inline assets/javascripts/card-form-customization.js:112]: The line above uses "Changes" instead of "Change". In the interests of consistency, should we make this `handleFontChange()` or rename `handleColorInputChanges()` to `handleColorInputChange()`?

**Review by AnasNaouchi — APPROVED** [review]

File: assets/javascripts/card-form-customization.js (hunk 2 above)
Comment (aashishgurung) [reply assets/javascripts/card-form-customization.js]: We should. Done. Thanks.

File: assets/javascripts/card-form-customization.js (hunk 3 above)
Comment (aashishgurung) [reply assets/javascripts/card-form-customization.js]: Changed it to single quotes

File: assets/javascripts/card-form-customization.js (hunk 5 above)
Comment (aashishgurung) [reply assets/javascripts/card-form-customization.js:112]: I didn't end `handleFontChange` with `s` because it touches only one field where as `handleColorInputChanges` touches multiple color input fields like text color, border color, label color and so on. Do you still think it would be better to end `handleFontChange` with `s`?

File: assets/javascripts/omise-embedded-card.js (hunk 4 above)
Comment (aashishgurung) [reply assets/javascripts/omise-embedded-card.js]: Done. Thanks.

File: tests/unit/includes/gateway/class-omise-payment-internetbanking-test.php [hunk 6]
```diff
@@ -8,7 +8,6 @@ public function setUp(): void
     {
         $this->sourceType = 'fpx';
         parent::setUp();
-        // require_once __DIR__ . '/../../../../includes/backends/class-omise-backend-fpx.php';
```
Comment (danfowler) [inline tests/unit/includes/gateway/class-omise-payment-internetbanking-test.php:11]: Why commented?

File: tests/unit/includes/gateway/bootstrap-test-setup.php [hunk 7]
```diff
@@ -18,6 +18,14 @@ function add_action() {}
         }
     }
 
+    /**
```
Comment (danfowler) [inline tests/unit/includes/gateway/bootstrap-test-setup.php:21]: Why this change?

File: includes/admin/views/omise-page-card-form-customization.php [hunk 8]
```diff
@@ -24,11 +25,19 @@
       <tr>
         <td class="text-bold" style="width: 250px;">Font Name</td>
         <td>
-          <select class="select-input" name="font[name]">
+          <select id="custom_sf_font_name" class="select-input" name="font[name]">
             <option value="Poppins">Poppins</option>
-            <option value="Circular" selected>Circular</option>
+            <option value="Circular">Circular</option>
```
Comment (danfowler) [inline includes/admin/views/omise-page-card-form-customization.php:30]: Why do we still have references to "Circular"?

File: assets/javascripts/card-form-customization.js (hunk 5 above)
Comment (danfowler) [reply assets/javascripts/card-form-customization.js:112]: I do prefer `handleFontChange`. I would also prefer `handleColorInputChange` as I'm not sure that the fact that it touches multiple fields is communicated by "changes" over "change", but we can discuss in a separate PR.

File: assets/javascripts/omise-embedded-card.js [hunk 9]
```diff
@@ -25,6 +25,12 @@ function showOmiseEmbeddedCardForm({
   }
   element.style.height = iframeElementHeight + 'px'
 
+  let fontName = font.name
+
+  if (font.name.trim() === OMISE_CUSTOM_FONT_OTHER && font.custom_name.trim() !== '') {
```
Comment (annatotest) [inline assets/javascripts/omise-embedded-card.js:30]: What if I select Other, but do not put anything to the Custom font field?

File: tests/unit/includes/gateway/class-omise-payment-internetbanking-test.php (hunk 6 above)
Comment (aashishgurung) [reply tests/unit/includes/gateway/class-omise-payment-internetbanking-test.php:11]: Don't remember. That's why I removed it.

File: tests/unit/includes/gateway/bootstrap-test-setup.php (hunk 7 above)
Comment (aashishgurung) [reply tests/unit/includes/gateway/bootstrap-test-setup.php:21]: We need to close Mockery after tests to clean up resources which we missed to include before.

https://docs.mockery.io/en/latest/reference/expectations.html#expectation-declarations

File: tests/unit/includes/gateway/class-omise-payment-internetbanking-test.php (hunk 6 above)
Comment (danfowler) [reply tests/unit/includes/gateway/class-omise-payment-internetbanking-test.php:11]: I typically prefer to remove altogether than just comment out.

File: tests/unit/includes/gateway/bootstrap-test-setup.php (hunk 7 above)
Comment (danfowler) [reply tests/unit/includes/gateway/bootstrap-test-setup.php:21]: It looks like the function was just moved from one place to another.

File: assets/javascripts/omise-embedded-card.js (hunk 9 above)
Comment (aashishgurung) [reply assets/javascripts/omise-embedded-card.js]: If that case ever occur then it will default to Poppins. We have added required attribute to the custom font name field if other is selected to avoid situation like this with HTML5 validation. You will see `Fill out this field` if you leave custom field empty.

File: includes/admin/views/omise-page-card-form-customization.php (hunk 8 above)
Comment (aashishgurung) [reply includes/admin/views/omise-page-card-form-customization.php:30]: I guess we never removed it from here.

File: tests/unit/includes/gateway/class-omise-payment-internetbanking-test.php (hunk 6 above)
Comment (danfowler) [reply tests/unit/includes/gateway/class-omise-payment-internetbanking-test.php:11]: My mistake.

File: includes/gateway/class-omise-payment-creditcard.php [hunk 10]
```diff
@@ -148,7 +148,7 @@ function init_form_fields() {
 					'css'         => Omise_Card_Image::get_css(),
 					'default'     => Omise_Card_Image::get_amex_default_display(),
 					'description' => wp_kses(
-						__( 'This only controls the icons displayed on the checkout page.<br />It is not related to card processing on Omise payment gateway.', 'omise' ),
+						__( 'This only controls the icons displayed on the checkout page.<br />It is not related to card processing on OPN payment gateway.', 'omise' ),
```
Comment (danfowler) [inline includes/gateway/class-omise-payment-creditcard.php:151]:
```suggestion
						__( 'This only controls the icons displayed on the checkout page.<br />It is not related to card processing on Opn Payments.', 'omise' ),
```

Comment (danfowler) [reply includes/gateway/class-omise-payment-creditcard.php:151]: Please verify, but I believe the current styling of the name is "Opn" not "OPN". Note: this was not always the case ¯\_(ツ)_/¯

File: tests/unit/includes/gateway/bootstrap-test-setup.php (hunk 7 above)
Comment (aashishgurung) [reply tests/unit/includes/gateway/bootstrap-test-setup.php:21]: It's because multiple classes are using Mockery. If a class uses Mockery then we have to close it with `tearDown`. Some classes does not have it because it's parent class includes it. `Bootstrap_Test_Setup` is a parent class to others. I will check the child classes and remove if it has been called unnecessarily.

File: includes/gateway/class-omise-payment-creditcard.php (hunk 10 above)
Comment (aashishgurung) [reply includes/gateway/class-omise-payment-creditcard.php]: I can see we have been using `Opn` instead of `OPN`. Thanks.

File: includes/gateway/class-omise-payment-creditcard.php [hunk 11]
```diff
@@ -15,7 +15,7 @@ public function __construct()
 		$this->has_fields         = true;
 		$this->method_title       = __( 'Opn Payments Credit / Debit Card', 'omise' );
 		$this->method_description = wp_kses(
-			__( 'Accept payment through <strong>Credit / Debit Card</strong> via Opn Payments payment gateway.', 'omise' ),
+			__( 'Accept payment through <strong>Credit / Debit Card</strong> via Opn payments.', 'omise' ),
```
Comment (danfowler) [inline includes/gateway/class-omise-payment-creditcard.php:18]:
```suggestion
			__( 'Accept payment through <strong>Credit / Debit Card</strong> via Opn Payments.', 'omise' ),
```

Comment (danfowler) [reply includes/gateway/class-omise-payment-creditcard.php:18]: capital P

**Review by danfowler — APPROVED** [review]

File: includes/gateway/traits/charge-request-builder-trait.php [hunk 12]
```diff
@@ -57,6 +57,7 @@ public function get_metadata($order_id, $additionalData = [])
 	 */
 	public function get_redirect_url($callback_url, $order_id, $order)
 	{
+		return 'https://opn.ooo';
```
Comment (danfowler) [inline includes/gateway/traits/charge-request-builder-trait.php:60]: Seems accidentally committed ☝🏽

Comment (aashishgurung) [reply includes/gateway/traits/charge-request-builder-trait.php]: Thanks.

**Review by annatotest — DISMISSED** [review]

**Review by danfowler — APPROVED** [review]

File: assets/javascripts/card-form-customization.js [hunk 13]
```diff
@@ -58,6 +58,28 @@ function getDesignFormValues() {
   return formValues;
 }
 
+function handleFontChange() {
+  const fontName = document.getElementById('custom_sf_font_name');
+
+  if (fontName.value === OMISE_CUSTOM_FONT_OTHER) {
+    document.getElementById('custom_sf_custom_font_name').style.display = null
```
Comment (danfowler) [inline assets/javascripts/card-form-customization.js:65]: Why not `fontName.style.display`?

Comment (aashishgurung) [reply assets/javascripts/card-form-customization.js:65]: These are two different elements. One has id of `custom_sf_font_name` and another has `custom_sf_custom_font_name`

Comment (danfowler) [reply assets/javascripts/card-form-customization.js:65]: oops 🤦🏽

**Review by danfowler — APPROVED** [review]

**Review by annatotest — APPROVED** [review]

**Review by AnasNaouchi — APPROVED** [review]
